### PR TITLE
Adding reference counting to thread_data

### PR DESCRIPTION
**Pull request by hkaiser**
This is the first step towards enabling scoped execution of threads

This PR now also has changes that are unrelated to the reference counting of threads but that have been necessary because of ASAN problems that started showing up. These changes are more pervasive than I was hoping for and are related to the initialization of the runtime, AGAS, parcelport, thread-manager, and io_pool classes. These changes are all in one commit (https://github.com/STEllAR-GROUP/hpx/pull/5441/commits/c2965e6bfee08b87582f46c6b0521b15f1cd6738). If needed, I could try separating those into their own PR that would have to be merged first. @msimberg please let me know how you would like to handle this.

### CODE REVIEW

**Comment by msimberg** [comment]
@hkaiser I planned to take a look at the `thread_pool_scheduler` test hanging/failling, but I see that it currently doesn't build. Should I wait a bit? Do you have a possible fix for it already?

**Comment by hkaiser** [comment]
> @hkaiser I planned to take a look at the `thread_pool_scheduler` test hanging/failling, but I see that it currently doesn't build. Should I wait a bit? Do you have a possible fix for it already?

Sorry for this. I'll fix it right away. You can go back one commit in the meantime (https://github.com/STEllAR-GROUP/hpx/pull/5441/commits/a3d371b2505fb655e14e04582171b2cf2b86914b), that should compile and will expose the same issue.

**Comment by msimberg** [comment]
@hkaiser this is keeping the `thread_pool_scheduler` test from exiting: https://github.com/STEllAR-GROUP/hpx/blob/877c2be8229a44810d0017ddd24ab8303b3570d2/libs/parallelism/executors/tests/unit/thread_pool_scheduler.cpp#L123. With the changes in this PR, is the recommendation to simply explicitly reset ids like that or is there another way? I see a `thread_id_noref_type`, but that looks like an internal type. To be clear, having a global `hpx::thread::id` like that is very test-specific and I'd hope noone does that in real code so I'd be fine with just explicitly requiring that it be reset if the plan is to keep an id like that alive for a longer time. The documentation for `hpx::this_thread::get_id()` and/or `hpx::thread::id` could probably do with a note about reference counting.

**Comment by hkaiser** [comment]
> @hkaiser this is keeping the `thread_pool_scheduler` test from exiting:
> 
> https://github.com/STEllAR-GROUP/hpx/blob/877c2be8229a44810d0017ddd24ab8303b3570d2/libs/parallelism/executors/tests/unit/thread_pool_scheduler.cpp#L123
> 
> . With the changes in this PR, is the recommendation to simply explicitly reset ids like that or is there another way? I see a `thread_id_noref_type`, but that looks like an internal type. To be clear, having a global `hpx::thread::id` like that is very test-specific and I'd hope noone does that in real code so I'd be fine with just explicitly requiring that it be reset if the plan is to keep an id like that alive for a longer time. The documentation for `hpx::this_thread::get_id()` and/or `hpx::thread::id` could probably do with a note about reference counting.

Thanks for diagnosing this! Not sure why it doesn't show up on all platforms, though... I think we can make the `hpx::thread::id` non-reference counting, however. That should resolve the issue.

**Comment by msimberg** [comment]
> > @hkaiser this is keeping the `thread_pool_scheduler` test from exiting:
> > https://github.com/STEllAR-GROUP/hpx/blob/877c2be8229a44810d0017ddd24ab8303b3570d2/libs/parallelism/executors/tests/unit/thread_pool_scheduler.cpp#L123
> > 
> > . With the changes in this PR, is the recommendation to simply explicitly reset ids like that or is there another way? I see a `thread_id_noref_type`, but that looks like an internal type. To be clear, having a global `hpx::thread::id` like that is very test-specific and I'd hope noone does that in real code so I'd be fine with just explicitly requiring that it be reset if the plan is to keep an id like that alive for a longer time. The documentation for `hpx::this_thread::get_id()` and/or `hpx::thread::id` could probably do with a note about reference counting.
> 
> Thanks for diagnosing this! Not sure why it doesn't show up on all platforms, though... I think we can make the `hpx::thread::id` non-reference counting, however. That should resolve the issue.

If by platform, you mean CI configuration, then most likely because that test is only built and run on those that are using C++17 or higher.

File: examples/balancing/hpx_thread_phase.cpp
Comment (msimberg) [inline examples/balancing/hpx_thread_phase.cpp:11]:
```suggestion
#include <hpx/local/functional.hpp>
```

File: examples/cancelable_action/cancelable_action/server/cancelable_action.hpp
Comment (msimberg) [inline examples/cancelable_action/cancelable_action/server/cancelable_action.hpp:54]: Just curious, why does this now require a lock rather than just being atomic?

File: libs/core/concurrency/include/hpx/concurrency/deque.hpp
Comment (msimberg) [inline libs/core/concurrency/include/hpx/concurrency/deque.hpp:29]: Are these changes related to the thread id changes. If yes, how?

File: libs/core/coroutines/include/hpx/coroutines/detail/coroutine_accessor.hpp
Comment (msimberg) [inline libs/core/coroutines/include/hpx/coroutines/detail/coroutine_accessor.hpp:41]: Not your change, but why is this static?

File: libs/core/schedulers/include/hpx/schedulers/thread_queue.hpp
Comment (msimberg) [inline libs/core/schedulers/include/hpx/schedulers/thread_queue.hpp:192]: Since we've talked about adding and not adding references before, how about an `enum class addref { yes, no }` or something similar in place of that bool?

File: libs/full/resiliency_distributed/include/hpx/resiliency_distributed/async_replay_distributed.hpp
Comment (msimberg) [inline libs/full/resiliency_distributed/include/hpx/resiliency_distributed/async_replay_distributed.hpp:69]: What's the reason for copying `this_` here rather than moving it? Or just capturing it directly as `[this_ = this->shared_from_this(), ...]`?

File: libs/core/coroutines/include/hpx/coroutines/thread_id_type.hpp
Comment (msimberg) [inline libs/core/coroutines/include/hpx/coroutines/thread_id_type.hpp:327]: I've nothing against this, but is this useful?

**Review by msimberg — COMMENTED** [review]
Thanks @hkaiser, this looks good to me. Just some minor comments, mainly out of curiousity.

File: examples/cancelable_action/cancelable_action/server/cancelable_action.hpp
Comment (hkaiser) [reply examples/cancelable_action/cancelable_action/server/cancelable_action.hpp:54]: I might be able to roll back this change now that `thread::id` is not reference counting anymore. The reference counted `thread::id` can't be an atomic.

File: libs/core/concurrency/include/hpx/concurrency/deque.hpp
Comment (hkaiser) [reply libs/core/concurrency/include/hpx/concurrency/deque.hpp:29]: No, it was a missing #include.

File: libs/core/coroutines/include/hpx/coroutines/detail/coroutine_accessor.hpp
Comment (hkaiser) [reply libs/core/coroutines/include/hpx/coroutines/detail/coroutine_accessor.hpp:41]: Because it does not need to access any members of that struct.

File: libs/core/schedulers/include/hpx/schedulers/thread_queue.hpp
Comment (hkaiser) [reply libs/core/schedulers/include/hpx/schedulers/thread_queue.hpp:192]: Good thinking!

File: libs/core/coroutines/include/hpx/coroutines/thread_id_type.hpp
Comment (hkaiser) [reply libs/core/coroutines/include/hpx/coroutines/thread_id_type.hpp:327]: It's needed if somebody wants to hold this type in a `std::set` or `std::map`, etc.

File: libs/full/resiliency_distributed/include/hpx/resiliency_distributed/async_replay_distributed.hpp
Comment (hkaiser) [reply libs/full/resiliency_distributed/include/hpx/resiliency_distributed/async_replay_distributed.hpp:69]: Good catch, that's a left-over from previous changes. I will roll this back.

File: libs/core/coroutines/include/hpx/coroutines/detail/coroutine_accessor.hpp
Comment (msimberg) [reply libs/core/coroutines/include/hpx/coroutines/detail/coroutine_accessor.hpp:39]: D'oh. I missed that it was inside a struct... Makes absolute sense.

File: examples/cancelable_action/cancelable_action/server/cancelable_action.hpp
Comment (msimberg) [reply examples/cancelable_action/cancelable_action/server/cancelable_action.hpp:54]: Was it not possible in the end or you just didn't get to it yet (or forget to push ;))?

Comment (hkaiser) [reply examples/cancelable_action/cancelable_action/server/cancelable_action.hpp:54]: It wasn't possible as even the non-ref-counted `thread;;id` ended up being not trivially copyable. I have not investigated whether this can be achieved yet, however.

Comment (msimberg) [reply examples/cancelable_action/cancelable_action/server/cancelable_action.hpp:54]: Ok, thanks. What makes the new non-ref-counted id different from the old non-ref-counted id on master? I imagined they'd be practically the same.

Comment (hkaiser) [reply examples/cancelable_action/cancelable_action/server/cancelable_action.hpp:54]: As said, I have not investigated yet. It could be the explicit move constructor/move assignment I added. We might be able to remove those again.

**Comment by hkaiser** [comment]
@msimberg this is finally ready for review. All what's left is to rebase it one more time. Please, if you have the time, have a closer look again.

File: libs/core/schedulers/include/hpx/schedulers/thread_queue.hpp
Comment (msimberg) [inline libs/core/schedulers/include/hpx/schedulers/thread_queue.hpp:260]: This assert seems odd to me, but I'm probably missing the bigger context. Why can we assume that `schedule_now` is always true here? The comment above makes it sound like it can still be false sometimes, but the assert doesn't match that.

File: libs/full/runtime_local/src/runtime_local.cpp
Comment (msimberg) [inline libs/full/runtime_local/src/runtime_local.cpp:370]: `init_tss` and `reinit_construct` don't seem to belong in `set_notification_policies`? Can they be moved to after the call to `set_notification_policies`?

**Comment by msimberg** [comment]
Sorry @hkaiser for taking so long with this. Two final comments, possibly non-issues.

> This PR now also has changes that are unrelated to the reference counting of threads but that have been necessary because of ASAN problems that started showing up. These changes are more pervasive than I was hoping for and are related to the initialization of the runtime, AGAS, parcelport, thread-manager, and io_pool classes. These changes are all in one commit ([c2965e6](https://github.com/STEllAR-GROUP/hpx/commit/c2965e6bfee08b87582f46c6b0521b15f1cd6738)). If needed, I could try separating those into their own PR that would have to be merged first. @msimberg please let me know how you would like to handle this.

As mentioned in last week's call, now that I understand why the "unrelated" changes are here, let's keep them in this PR.

File: libs/core/schedulers/include/hpx/schedulers/thread_queue.hpp
Comment (hkaiser) [reply libs/core/schedulers/include/hpx/schedulers/thread_queue.hpp:260]: In general, the scheduling system now keeps threads alive if they are `pending` (sitting in a queue), `active`/running (the scheduling loop holds them), are `suspended` (there is a reference count sitting on the thread's stack created before the actual suspension), or explicitly by the user. 

Therefore, staged tasks must have their initial state set to `pending`. They are converted to real threads without user intervention by the scheduler, thus the only way to keep them alive (keep the reference count up) is to schedule them (add them to a queue) once they have been created.

This assert should be always true because of this: https://github.com/STEllAR-GROUP/hpx/pull/5441/files/c2965e6bfee08b87582f46c6b0521b15f1cd6738#diff-34057d72d801cb0ee81de2f396edd2f1de7bd0d6309445a06df1d1e58630de8eR733. I should however turn the second assert into an exception to enforce the first.

File: libs/full/runtime_local/src/runtime_local.cpp
Comment (hkaiser) [reply libs/full/runtime_local/src/runtime_local.cpp:370]: Yes, definitely.

File: libs/core/schedulers/include/hpx/schedulers/thread_queue.hpp
Comment (msimberg) [reply libs/core/schedulers/include/hpx/schedulers/thread_queue.hpp:260]: Do we then even need the `staged` state anymore?

Comment (hkaiser) [reply libs/core/schedulers/include/hpx/schedulers/thread_queue.hpp:260]: The staged state represents a thread to be delay-created only once the scheduler runs out of work,  so this question is an orthogonal one. The new limitation really just requires for staged threads to be scheduled right away once they are created eventually. I think this is what I would expect anyways. What use is a staged task if it should not be run once turned into a 'real' thread?

Comment (msimberg) [reply libs/core/schedulers/include/hpx/schedulers/thread_queue.hpp:260]: Fully agreed. The `run_now` flag represents whether or not a thread should start as `staged` or be scheduled immediately, not the `staged` state. As far as I can tell, and this seems to have been true before this PR as well, the only places we use `thread_schedule_state::staged` is to query the sizes of the staged tasks queues. Neither `thread_init_data` nor `thread_data` should ever be in the `staged` state (there's an `initial_state_` in `thread_init_data` but that's for when the thread actually is scheduled).

With that in mind should we also add an assertion that `initial_state_` is never `staged` to the `thread_init_data` constructors: https://github.com/STEllAR-GROUP/hpx/blob/fa6e4df704bf5f81ddc0f4e09716bd11ff8724c8/libs/core/threading_base/include/hpx/threading_base/thread_init_data.hpp#L29-L133? In addition to the assertions that you've already added and proposed, that is.

**Comment by hkaiser** [comment]
> @hkaiser this is keeping the `thread_pool_scheduler` test from exiting:
> 
> https://github.com/STEllAR-GROUP/hpx/blob/877c2be8229a44810d0017ddd24ab8303b3570d2/libs/parallelism/executors/tests/unit/thread_pool_scheduler.cpp#L123
> 
> . With the changes in this PR, is the recommendation to simply explicitly reset ids like that or is there another way? I see a `thread_id_noref_type`, but that looks like an internal type. To be clear, having a global `hpx::thread::id` like that is very test-specific and I'd hope noone does that in real code so I'd be fine with just explicitly requiring that it be reset if the plan is to keep an id like that alive for a longer time. The documentation for `hpx::this_thread::get_id()` and/or `hpx::thread::id` could probably do with a note about reference counting.

`hpx::thread::id` is non-reference counting now. I think that there are no changes for the user.

File: libs/core/schedulers/include/hpx/schedulers/thread_queue.hpp
Comment (hkaiser) [reply libs/core/schedulers/include/hpx/schedulers/thread_queue.hpp:260]: I agree that the initial state should not be `staged`, ever. It's an internal state users shouldn't have to worry about. I will add an error check for this.

**Review by aurianer — APPROVED** [review]
Thanks!

**Comment by hkaiser** [comment]
> Uh, sorry @hkaiser. I merged #5377 which conflicts with this. It may be easiest to ignore the changes to the `set_thread_state` files that I did in #5377, and I can try to redo them on top of your changes. Basically they boiled down to separating the timed and non-timed functionality to separate headers to avoid pulling in Asio in as many places. In any case, if you need a hand with this let me know.

@msimberg I rebased and tried to re-apply your changes. Could you please verify I didn't mess anything up?

**Comment by msimberg** [comment]
> @msimberg I rebased and tried to re-apply your changes. Could you please verify I didn't mess anything up?

The changes look very sane, thank you! The move of the implementations to cpp files is also two steps better than the changes I had in #5377. :+1:

**Review by msimberg — DISMISSED** [review]
Thanks @hkaiser. Let's go ahead with this if it comes back clean (we might have to wait for at least the Piz Daint builders to be back in action for that though...).

**Comment by msimberg** [comment]
retest cscs

**Comment by hkaiser** [comment]
This should be good to go once the tests have cycled.

**Comment by hkaiser** [comment]
During the PMC meeting today we agreed that this is good to go.

**Pull request merged**
